src/vma/dev/cq_mgr.cpp

```diff
@@ -526,8 +526,7 @@ mem_buf_desc_t* cq_mgr::process_cq_element_rx(vma_ibv_wc* p_wce)
 	// Get related mem_buf_desc pointer from the wr_id
 	mem_buf_desc_t* p_mem_buf_desc = (mem_buf_desc_t*)(uintptr_t)p_wce->wr_id;
 
-	bool bad_wce = (p_wce->status != IBV_WC_SUCCESS) ||
-			(m_b_is_rx_csum_on && !vma_wc_rx_csum_ok(*p_wce));
+	bool bad_wce = p_wce->status != IBV_WC_SUCCESS;
```

Comment:
1. You are right,  HW CSUM can fail not only when the CSUM value is inncorrect,  but also when using some header options which are not supported by the HW.
Hence, we have to check it before we want to drop the packet.

2.  HW CSUM includes both IP and TCP/UDP csum as defiened in verbs_extra.h:  
old MLNX_OFED verbs (2.1 and older) :  
```
#ifdef DEFINED_IBV_DEVICE_RAW_IP_CSUM
#define vma_is_rx_csum_supported(attr)		((attr).device_cap_flags & (IBV_DEVICE_RAW_IP_CSUM | IBV_DEVICE_UD_IP_CSUM))
#define vma_wc_rx_csum_ok(wc)			(vma_wc_flags(wc) & IBV_WC_IP_CSUM_OK)
#else
```
From ibv_poll_cq manual :   

> IBV_WC_IP_CSUM_OK  TCP/UDP checksum over IPv4 and IPv4 header checksum are verified.

More info : http://permalink.gmane.org/gmane.linux.drivers.rdma/28565
else :    
```
#ifdef DEFINED_IBV_EXP_DEVICE_RX_CSUM_L4_PKT
#define vma_wc_rx_csum_ok(wc)			((vma_wc_flags(wc) & IBV_EXP_L3_RX_CSUM_OK) && (vma_wc_flags(wc) & IBV_EXP_L4_RX_CSUM_OK))
#else
#ifdef DEFINED_IBV_EXP_DEVICE_RX_CSUM_TCP_UDP_PKT
#define vma_wc_rx_csum_ok(wc)			((vma_wc_flags(wc) & IBV_EXP_WC_RX_IP_CSUM_OK) && (vma_wc_flags(wc) & IBV_EXP_WC_RX_TCP_UDP_CSUM_OK))
```